### PR TITLE
simpler 'cpu_scatter_gather_base_kernel'

### PR DESCRIPTION
**Pull request by nikitaved**
Simplifies `cpu_scatter_gather_base_kernel` to accept only binary operations and spares them from doing redundant checks.
CC @v0dro

### CODE REVIEW

**Comment by nikitaved** [comment]
Interesting, while `scatter` shows (sometimes even significant) performance gain, `gather` has some reductions (5-10%). I will try to investigate whether the cause is the loop-callback vs callback-loop order.

**Comment by nikitaved** [comment]
@ngimel , I did some additional loop optimizations and I observed some significant performance gains in the following "dense indexing" scenario:
```python
import torch
import numpy
from IPython import get_ipython

numpy.random.seed(13)
torch.manual_seed(13)

ipython = get_ipython()

torch.set_num_threads(1)

Ms=256
Ns=512
dim = 0
top_power = 5

for pM in range(top_power):
    M = Ms * (2 ** pM)
    for pN in range(top_power):
        N = Ns * (2 ** pN)
        input_one = torch.rand(M, N)
        index = torch.tensor(numpy.random.randint(0, M, (M, N)))
        print(f"Problem size (MxN): {M}x{N}")
        ipython.magic("timeit torch.gather(input_one, dim, index)")
```
### Before
```
Problem size (MxN): 256x512
762 µs ± 5.17 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 256x1024
2.12 ms ± 33.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Problem size (MxN): 256x2048
3.89 ms ± 9.65 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Problem size (MxN): 256x4096
7.79 ms ± 10.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Problem size (MxN): 256x8192
21.2 ms ± 2.96 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Problem size (MxN): 512x512
979 µs ± 1.33 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 512x1024
2.98 ms ± 3.12 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Problem size (MxN): 512x2048
8.18 ms ± 58 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Problem size (MxN): 512x4096
21.3 ms ± 214 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Problem size (MxN): 512x8192
44.5 ms ± 1.04 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
Problem size (MxN): 1024x512
3.52 ms ± 7.07 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Problem size (MxN): 1024x1024
7.74 ms ± 33.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Problem size (MxN): 1024x2048
15.2 ms ± 481 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Problem size (MxN): 1024x4096
49.8 ms ± 344 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Problem size (MxN): 1024x8192
122 ms ± 334 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Problem size (MxN): 2048x512
7.04 ms ± 7.03 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Problem size (MxN): 2048x1024
19 ms ± 53.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Problem size (MxN): 2048x2048
50.8 ms ± 382 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Problem size (MxN): 2048x4096
172 ms ± 45 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Problem size (MxN): 2048x8192
363 ms ± 472 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
Problem size (MxN): 4096x512
20.3 ms ± 15.9 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Problem size (MxN): 4096x1024
71.6 ms ± 282 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Problem size (MxN): 4096x2048
179 ms ± 1.19 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
Problem size (MxN): 4096x4096
379 ms ± 622 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
Problem size (MxN): 4096x8192
803 ms ± 306 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

### After
```
Problem size (MxN): 256x512                                                                                                                       
145 µs ± 1.1 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)                                                                           
Problem size (MxN): 256x1024                                                                                                                      
441 µs ± 410 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)                                                                            
Problem size (MxN): 256x2048                                                                                                                      
963 µs ± 1.09 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)                                                                           
Problem size (MxN): 256x4096                                                                                                                      
2.87 ms ± 6.12 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)                                                                           
Problem size (MxN): 256x8192                                                                                                                      
11.9 ms ± 5.67 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)                                                                           
Problem size (MxN): 512x512                                                                                                                       
328 µs ± 709 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)                                                                            
Problem size (MxN): 512x1024                                                                                                                      
998 µs ± 563 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)                                                                            
Problem size (MxN): 512x2048                                                                                                                      
2.53 ms ± 3.41 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Problem size (MxN): 512x4096
11.7 ms ± 136 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Problem size (MxN): 512x8192
33.5 ms ± 31.6 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Problem size (MxN): 1024x512
4.31 ms ± 6.15 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Problem size (MxN): 1024x1024
3.03 ms ± 3.21 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Problem size (MxN): 1024x2048
11.2 ms ± 19 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Problem size (MxN): 1024x4096
31 ms ± 130 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Problem size (MxN): 1024x8192
85 ms ± 470 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Problem size (MxN): 2048x512
7.74 ms ± 30.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Problem size (MxN): 2048x1024
20.4 ms ± 27.9 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Problem size (MxN): 2048x2048
28.6 ms ± 25.2 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Problem size (MxN): 2048x4096
79.6 ms ± 327 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Problem size (MxN): 2048x8192
166 ms ± 346 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Problem size (MxN): 4096x512
22.4 ms ± 22.1 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Problem size (MxN): 4096x1024
72.1 ms ± 248 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Problem size (MxN): 4096x2048
179 ms ± 103 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Problem size (MxN): 4096x4096
158 ms ± 35 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Problem size (MxN): 4096x8192
324 ms ± 487 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

```

The rest is about 2-5% faster or very similar.

**Comment by ngimel** [comment]
This is impressive! Do the gains hold for both scatter and gather? I'll review and merge soon, thank you!

**Comment by nikitaved** [comment]
scatter is less impressive, maybe by a factor of about 2x in the "dense indexing" test scenario. I am not sure that the heuristic that I use for this very simple loop optimization is optimal, however, need some expert's opinion.

**Comment by nikitaved** [comment]
Ah, I also observed the effect of the indexed dimension. I will update my heuristic to take into account the contiguous structure...

**Comment by ngimel** [comment]
This is a cool simplification! Indexing dimension does matter. Rough heuristics is innermost loop should be over fastest varying (last) dimension. This is complicated by the fact that when you are scattering/gathering over last dimension, you accesses are still irregular, but that was always the case, so that alone should not lead to perf regressions. So I would hope that for operations on the last dim current structure should be ok (always calling `_cpu_scatter_gather_dim_loop`), and for other dimensions the heuristics that you had here previously? 
Can we have a recap of the perf changes due to this PR? gather along dim=0 can be made faster with the previous heuristics, right? scatter along dim=0 is approximately unchanged, which is still fine. What about dim=1 results?

**Comment by nikitaved** [comment]
@ngimel , thank you! I adopted your suggestions. So, `dim=0` is faster for both gather(2-5x) and scatter(about 2x) with "dense indexing". `dim=1` is unchanged.

**Comment by ngimel** [comment]
Cool! Can you also please make sure that cases from here https://github.com/pytorch/pytorch/pull/32425#issuecomment-576787143 don't regress, and we should be good to go.

**Comment by nikitaved** [comment]
No, these cases do not regress. What do you think about the case when the indexed dimension is in the outer most loop? Shall I also try putting the loop over the other dimensions in a separate code block to see whether this might have any positive effect?

**Comment by ngimel** [comment]
It might help, but tbh I don't think it's worth added code complexity. I'll try landing this as is.
